File: BackEnd/PHP/laravelshop/app/Modules/Matching/Services/MatchingService.php
<?php

namespace App\Modules\Matching\Services;

use App\Common\Exceptions\BusinessException;
use App\Models\Product;
use App\Models\ProductMapping;
use App\Models\User;
use Symfony\Component\HttpFoundation\Response;

class MatchingService
{
    public function pending(User $actor, int $page, int $limit): array
    {
        $this->assertAdmin($actor);
        $result = ProductMapping::query()->where('status', 'PENDING')->orderByDesc('id')->paginate($limit, ['*'], 'page', $page);

        return [
            'items' => $result->getCollection()->map(fn (ProductMapping $mapping): array => $this->serialize($mapping))->all(),
            'pagination' => $this->pagination($result),
        ];
    }

    public function approve(User $actor, int $id, int $productId): array
    {
        $this->assertAdmin($actor);
        $mapping = ProductMapping::query()->find($id);
        $product = Product::query()->find($productId);
        if (! $mapping) {
            throw new BusinessException('매핑 대상을 찾을 수 없습니다.', 'MAPPING_NOT_FOUND', Response::HTTP_NOT_FOUND);
        }
        if (! $product) {
            throw new BusinessException('상품을 찾을 수 없습니다.', 'PRODUCT_NOT_FOUND', Response::HTTP_NOT_FOUND);
        }

        $mapping->forceFill(['product_id' => $productId, 'status' => 'APPROVED', 'reason' => null])->save();

        return ['message' => '매핑을 승인했습니다.'];
    }

    public function reject(User $actor, int $id, string $reason): array
    {
        $this->assertAdmin($actor);
        $mapping = ProductMapping::query()->find($id);
        if (! $mapping) {
            throw new BusinessException('매핑 대상을 찾을 수 없습니다.', 'MAPPING_NOT_FOUND', Response::HTTP_NOT_FOUND);
        }

        $mapping->forceFill(['status' => 'REJECTED', 'reason' => $reason])->save();

        return ['message' => '매핑을 거절했습니다.'];
    }

    public function autoMatch(User $actor): array
    {
        $this->assertAdmin($actor);
        $pending = ProductMapping::query()->where('status', 'PENDING')->get();
        $matched = 0;

        foreach ($pending as $mapping) {
            $product = Product::query()->where('name', 'like', '%' . $mapping->source_name . '%')->first();
            if ($product) {
                $mapping->forceFill(['product_id' => $product->id, 'status' => 'APPROVED', 'reason' => 'AUTO_MATCHED'])->save();
                $matched++;
            }
        }

        return [
            'matchedCount' => $matched,
            'pendingCount' => ProductMapping::query()->where('status', 'PENDING')->count(),
        ];
    }

    public function stats(User $actor): array
    {
        $this->assertAdmin($actor);

        return [
            'pending' => ProductMapping::query()->where('status', 'PENDING')->count(),
            'approved' => ProductMapping::query()->where('status', 'APPROVED')->count(),
            'rejected' => ProductMapping::query()->where('status', 'REJECTED')->count(),
        ];
    }

    private function serialize(ProductMapping $mapping): array
    {
        return [
            'id' => $mapping->id,
            'sourceName' => $mapping->source_name,
            'productId' => $mapping->product_id,
            'status' => $mapping->status,
            'reason' => $mapping->reason,
            'createdAt' => optional($mapping->created_at)?->toISOString(),
            'updatedAt' => optional($mapping->updated_at)?->toISOString(),
        ];
    }

    private function pagination($result): array
    {
        return [
            'page' => $result->currentPage(),
            'limit' => $result->perPage(),
            'total' => $result->total(),
            'totalPages' => $result->lastPage(),
        ];
    }

    private function assertAdmin(User $user): void
    {
        if ($user->role !== 'ADMIN') {
            throw new BusinessException('관리자 권한이 필요합니다.', 'FORBIDDEN', Response::HTTP_FORBIDDEN);
        }
    }
}

File: BackEnd/Java/java-spring-maven-jpa-postgresql/src/main/java/com/pbshop/java/spring/maven/jpa/postgresql/matching/MatchingService.java
package com.pbshop.java.spring.maven.jpa.postgresql.matching;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.pbshop.java.spring.maven.jpa.postgresql.auth.security.AuthenticatedUserPrincipal;
import com.pbshop.java.spring.maven.jpa.postgresql.common.exception.BusinessException;
import com.pbshop.java.spring.maven.jpa.postgresql.common.exception.ErrorCode;
import com.pbshop.java.spring.maven.jpa.postgresql.product.Product;
import com.pbshop.java.spring.maven.jpa.postgresql.product.ProductRepository;

@Service
@Transactional
public class MatchingService {
    private final ProductMappingRepository productMappingRepository;
    private final ProductRepository productRepository;
    public MatchingService(ProductMappingRepository productMappingRepository, ProductRepository productRepository) {
        this.productMappingRepository = productMappingRepository;
        this.productRepository = productRepository;
    }

    @Transactional(readOnly = true)
    public Map<String, Object> pending(AuthenticatedUserPrincipal principal) {
        requireAdmin(principal);
        List<Map<String, Object>> items = productMappingRepository.findByStatusOrderByIdDesc("PENDING").stream().map(this::toResponse).toList();
        return Map.of("items", items, "pagination", Map.of("page", 1, "limit", items.size() == 0 ? 20 : items.size(), "total", items.size(), "totalPages", items.isEmpty() ? 0 : 1));
    }

    public Map<String, Object> approve(AuthenticatedUserPrincipal principal, Long id, Long productId) {
        requireAdmin(principal);
        ProductMapping mapping = requireMapping(id);
        Product product = productRepository.findById(productId)
                .orElseThrow(() -> new BusinessException(ErrorCode.NOT_FOUND, "상품을 찾을 수 없습니다."));
        mapping.setProduct(product);
        mapping.setStatus("APPROVED");
        mapping.setReason(null);
        return Map.of("message", "매핑을 승인했습니다.");
    }

    public Map<String, Object> reject(AuthenticatedUserPrincipal principal, Long id, String reason) {
        requireAdmin(principal);
        ProductMapping mapping = requireMapping(id);
        mapping.setStatus("REJECTED");
        mapping.setReason(reason);
        return Map.of("message", "매핑을 거절했습니다.");
    }

    public Map<String, Object> autoMatch(AuthenticatedUserPrincipal principal) {
        requireAdmin(principal);
        int matched = 0;
        for (ProductMapping mapping : productMappingRepository.findByStatusOrderByIdDesc("PENDING")) {
            Product product = productRepository.findAll().stream()
                    .filter(item -> item.getName().contains(mapping.getSourceName()))
                    .findFirst().orElse(null);
            if (product != null) {
                mapping.setProduct(product);
                mapping.setStatus("APPROVED");
                mapping.setReason("AUTO_MATCHED");
                matched++;
            }
        }
        return Map.of("matchedCount", matched, "pendingCount", productMappingRepository.countByStatus("PENDING"));
    }

    @Transactional(readOnly = true)
    public Map<String, Object> stats(AuthenticatedUserPrincipal principal) {
        requireAdmin(principal);
        return Map.of(
                "pending", productMappingRepository.countByStatus("PENDING"),
                "approved", productMappingRepository.countByStatus("APPROVED"),
                "rejected", productMappingRepository.countByStatus("REJECTED"));
    }

    private ProductMapping requireMapping(Long id) {
        return productMappingRepository.findById(id)
                .orElseThrow(() -> new BusinessException(ErrorCode.NOT_FOUND, "매핑 대상을 찾을 수 없습니다."));
    }

    private Map<String, Object> toResponse(ProductMapping mapping) {
        Map<String, Object> item = new LinkedHashMap<>();
        item.put("id", mapping.getId());
        item.put("sourceName", mapping.getSourceName());
        item.put("productId", mapping.getProduct() == null ? null : mapping.getProduct().getId());
        item.put("status", mapping.getStatus());
        item.put("reason", mapping.getReason() == null ? "" : mapping.getReason());
        item.put("createdAt", mapping.getCreatedAt() == null ? null : mapping.getCreatedAt().toString());
        item.put("updatedAt", mapping.getUpdatedAt() == null ? null : mapping.getUpdatedAt().toString());
        return item;
    }

    private void requireAdmin(AuthenticatedUserPrincipal principal) {
        if (principal == null) {
            throw new BusinessException(ErrorCode.UNAUTHORIZED, "인증이 필요합니다.");
        }
        if (!"ADMIN".equalsIgnoreCase(principal.role())) {
            throw new BusinessException(ErrorCode.FORBIDDEN, "관리자 권한이 필요합니다.");
        }
    }
}
